compiler/rustc_middle/src/ty/diagnostics.rs

```diff
@@ -271,6 +271,20 @@ pub fn suggest_constraining_type_params<'a>(
             }
         }
 
+        // in the scenario like impl has stricter requirements than trait,
+        // we should not suggest restrict bound on the impl, here we double check
+        // the whether the param already has the constraint by checking `def_id`
+        let bound_trait_defs: Vec<DefId> = generics
+            .bounds_for_param(param.def_id)
+            .flat_map(|bound| {
+                bound.bounds.iter().flat_map(|b| b.trait_ref().and_then(|t| t.trait_def_id()))
+            })
+            .collect();
+        constraints = constraints
+            .into_iter()
+            .filter(|(_, def_id)| def_id.map_or(true, |def| !bound_trait_defs.contains(&def)))
+            .collect();
```

Comment:
Suggestion: can this use [`Vec::retain`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.retain)?